File: BW8AGU/VB/Ch08/BW8AGU.Samples.VB/BW8AGU.Samples.VB/Common/PasswordVaultHelper.vb
Imports Windows.Security.Credentials

Public Class PasswordVaultHelper
    Private Shared vault As PasswordVault = New Windows.Security.Credentials.PasswordVault()

    Public Shared Function LoadPasswordVault() As IList(Of PasswordCredential)
        'Load all credentials
        Dim creds As IReadOnlyList(Of PasswordCredential) = vault.RetrieveAll()

        Return creds.ToList()
    End Function

    Public Shared Function CheckPassword(username As String, password As String) As Boolean
        Dim res As Boolean = False

        Try
            Dim creds As IReadOnlyList(Of PasswordCredential) = vault.FindAllByUserName(username)

            For Each cred As PasswordCredential In creds
                cred.RetrievePassword()
                res = cred.Password = password
            Next
            'do nothing
        Catch
        End Try

        Return res
    End Function

    Public Shared Sub SavePasswordToVault(username As String, password As String)
        LoadPasswordVault()

        Dim cred As New PasswordCredential("BW8AGU", username, password)

        vault.Add(cred)
    End Sub
End Class

File: BW8AGU/C#/Ch08/BW8AGU.Samples.CSharp/Common/PasswordVaultHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.Security.Credentials;

namespace BW8AGU.Samples.CSharp.Common
{
    public class PasswordVaultHelper
    {
        private static PasswordVault vault =
            new Windows.Security.Credentials.PasswordVault();

        public static IList<PasswordCredential> LoadPasswordVault()
        {
            //Load all credentials
            IReadOnlyList<PasswordCredential> creds = vault.RetrieveAll();

            return creds.ToList<PasswordCredential>();
        }

        public static bool CheckPassword(string username, string password)
        {
            bool res = false;

            try
            {
                IReadOnlyList<PasswordCredential> creds = vault.FindAllByUserName(username);

                foreach (PasswordCredential cred in creds)
                {
                    cred.RetrievePassword();
                    res = cred.Password == password;
                }
            }
            catch
            {
                //do nothing
            }

            return res;
        }

        public static void SavePasswordToVault(string username, string password)
        {
            LoadPasswordVault();

            PasswordCredential cred = new PasswordCredential("BW8AGU", username, password);

            vault.Add(cred);
        }
    }
}
